docs/USAGE-GUIDE.md

```diff
@@ -1777,6 +1777,210 @@ handle. The handle is considered invalid after **s2n_connection_free** is used.
+
+**s2n_psk_set_identity** sets the PSK identity for a given PSK. The PSK identity is a unique identifier for the pre-shared secret. This identity is transmitted over the network unencrypted and is a non-secret value therefore do not include any confidential information in the identity.
+
+**s2n_psk_set_secret** sets the secret value for a given PSK. Note that the secret value is copied into s2n-tls memory and the user is responsible for freeing the memory associated with the **secret** input.
+
+**s2n_psk_set_hmac** sets the PSK hash algorithm for a given PSK.
+The supported PSK hash algorithms are listed in the **s2n_psk_hmac** enum. This API overrides the default PSK hash algorithm value of **S2N_PSK_HMAC_SHA256**. Use this API if you wish to select a different cipher suite, as the selected cipher suite must have the same hash algorithm as the PSK.
+
+**s2n_connection_append_psk** appends the PSK to the connection. Both server and client should call this API to add PSKs to their connection. Note that the order this API is called matters, as PSKs that are appended first will be more preferred than PSKs appended last. This API sets up the connection to be later used by the server PSK extension to select a PSK for the connection. If a PSK with a duplicate identity is found, an error is returned and the PSK is not added to the list. If the connection already contains a list of PSKs **s2n_connection_append_psk** will verify that the PSK mode being set matches that of the list. If the PSK modes differs, an error is returned. If there is no existing list of PSKs present on the connection (the input PSK is the first PSK to be added to the list) the PSK mode of the list is set to the input PSK mode. The PSK object is copied to the connection and the user is responsible for freeing the memory associated with the PSK object.
+
+**s2n_config_set_psk_mode** sets the PSK mode on the config object. The PSK mode represents how the pre-shared key was generated and shared between the server and the client. The enum **s2n_psk_mode** lists the supported PSK modes. Use this API prior to the ClientHello to set the PSK mode for the connection. This API can also be used to switch between the different PSK modes.
+
+**s2n_connection_set_psk_mode** sets the PSK mode on the connection object. The enum **s2n_psk_mode** lists the supported PSK modes. This API overrides the PSK mode set on config for this connection. If the connection already contains a list of PSKs **s2n_connection_set_psk_mode** verifies that the current PSK mode matches that of the list. If the PSK mode being set differs from the existing list, an error is returned. If there is no existing list of PSKs present on the connection, the PSK mode of the list is set to the input PSK mode.
+
+### Selecting a Pre-Shared Key
+
```

Comment:
Can we include a discussion of the case where the server has no PSKs set when this callback is called, including how how s2n_connection_set_psk_mode factors in? That will likely be the main use case for this callback, and it's not obvious how to handle it.